### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by Writhe**
There you have it. Multi-team support is probably quite wobbly - for instance, it should check if addressee.teamSlackID === sender.teamSlackID, otherwise there is a (very small) chance that someone from another org gets an undeserved kudo.

### CODE REVIEW

File: internal/misc/bullets.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package misc
+
+import "math/rand"
+
+var bullets = []string{":green_heart:", ":heart:", ":blue_heart:", ":yellow_heart:", ":purple_heart:", ":orange_heart:", ":black_heart:"}
```
Comment (papisz) [inline internal/misc/bullets.go:5]: perhaps it would be good to have this in a config file

File: internal/misc/exclamations.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package misc
+
+import "math/rand"
+
+var exclamations = []string{"What an absolute unit!", "Bravo! :clap:", "What a trooper!", ":tada:", "Look at you go!", "So classy! :face_with_monocle:"}
```
Comment (papisz) [inline internal/misc/exclamations.go:5]: this might also be configurable

File: internal/server/server.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+// CommandServer - Kudosaurus' Slack slash-command server
+type CommandServer struct {
+	store  kudosaurus.Store
+	api    *slack.Client
+	config CommandServerConfig
+	logger *log.Logger
+}
+
+func (s *CommandServer) getUserName(id string) (string, bool) {
```
Comment (papisz) [inline internal/server/server.go:32]: in general it is a bad idea to have one struct which handles everything. It's better to make small structures, which implement small interfaces and bind it all together in `main.go`. Here you have those commands scattered around a couple files and it makes the code hard to follow.

Here's an example which I googled, but it should give you the idea:

```go
type PoemStorage interface {
	Load(string) []byte
	Save(string, []byte)
}
func NewPoem(ps PoemStorage) *Poem {
	return &Poem{
		storage: ps
	}
}
```
And then the main:
```go
func main() {
	storage := NewNapkin()
	poem := NewPoem(storage)  // wired up.
}
```

Full article for reference: https://appliedgo.net/di/

File: internal/sqlitestore/person.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// GetFullData - gets full kudosaurus.Person data from the wrapper
+func (p Person) GetFullData(db *gorm.DB) kudosaurus.Person {
+	var mappedKudos []kudosaurus.Kudo
+	var mappedIssuedKudos []kudosaurus.Kudo
+
+	kudos := p.GetKudos(db)
+	issuedKudos := p.GetIssuedKudos(db)
+	kudosLeft := 10 - len(issuedKudos)
```
Comment (papisz) [inline internal/sqlitestore/person.go:47]: `10` sounds like something that should be configurable :)

File: internal/sqlitestore/store.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+	"github.com/jinzhu/gorm"
+	"github.com/writhe/kudosaurus"
+
+	// NOTE: Needed for Sqlite3 support
+	_ "github.com/jinzhu/gorm/dialects/sqlite"
+)
+
+// SqliteStore - implements interface DataSource
+type SqliteStore struct {
```
Comment (papisz) [inline internal/sqlitestore/store.go:17]: I like the fact that you've separated SQL functions from rest of the app. That's a good practice.

File: kudosaurus.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
+package kudosaurus
```
Comment (papisz) [inline kudosaurus.go:1]: that's how the code should be structured 👍 (all entities exactly where they should be)

**Review by papisz — COMMENTED** [review]
There are parts which are done very well, but `CommandServer` is too big in my opinion and should be split into many interfaces, bound together in main.go file with dependency injection. Right now it has too many responsibilities and mixes levels of abstractions.

File: internal/misc/bullets.go (hunk 1 above)
Comment (Writhe) [reply internal/misc/bullets.go:5]: Definitely. It's something I _know_ how to do, so I decided to focus on more difficult aspects. But, yeah - good catch.

File: internal/misc/exclamations.go (hunk 2 above)
Comment (Writhe) [reply internal/misc/exclamations.go:5]: Ditto.

File: internal/server/server.go (hunk 3 above)
Comment (Writhe) [reply internal/server/server.go:32]: Yeah, I figured this is becoming a bit of a god object, but I know nothing about proper code structure in Go. Will check out the article, thanks!

File: internal/sqlitestore/person.go (hunk 4 above)
Comment (Writhe) [reply internal/sqlitestore/person.go:47]: Good catch - there is a config variable, but I must've forgotten to use it, here. 🤦

File: internal/sqlitestore/store.go (hunk 5 above)
Comment (Writhe) [reply internal/sqlitestore/store.go:17]: The idea was to be able to swap out Sqlite for another data source.

File: kudosaurus.go (hunk 6 above)
Comment (Writhe) [reply kudosaurus.go:1]: 😊 

I think that's something TheMacies suggested.

File: kudosaurus.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// Settings - runtime settings for a team
+type Settings struct {
+	TeamID          string
+	TargetChannelID string
+}
+
+// Store - stores data
+type Store interface {
```
Comment (KNiepok) [inline kudosaurus.go:38]: I think it might be a good idea to separate it into few interfaces (one per each entity if possible)
Also PublishKudos() could rather be updateKudos() (from perspective of a storage layer, context of business case is irrelevant)

File: internal/server/actions.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+		s.closeWithError(errors.New("this is not your kudo to delete"), http.StatusUnauthorized, res)
+		return
+	}
+
+	if err := s.store.RemoveKudo(id); err != nil {
+		s.closeWithError(err, http.StatusInternalServerError, res)
+		return
+	}
+
+	res.WriteHeader(http.StatusOK)
```
Comment (KNiepok) [inline internal/server/actions.go:44]: WriteHeader is really only useful for setting any status besides 200 OK, as this one is assigned by default.
So unless you changed it at any earlier stage, these `res.WriteHeader(http.StatusOK)` can be removed

File: kudosaurus.go (hunk 7 above)
Comment (TheMacies) [inline kudosaurus.go:38]: Seems to me that this interface could be at least composed of smaller interfaces like
`PersonStore`,`KudoStore`,`TeamsStore`,`AdminStore`. This would increase readability and it's a good practice in general.

File: kudosaurus.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	PutPerson(id string, name string, teamID string, teamName string) (Person, error)
```
Comment (TheMacies) [inline kudosaurus.go:39]: I do not see `context.Context` being passed anywhere in here. Passing ctx all the way down from HTTP handler to database implementation allows early cancelation of long running tasks and is good practice in general. You could have some important information injected into the context.

File: kudosaurus.go (hunk 7 above)
Comment (Writhe) [reply kudosaurus.go:38]: Yeah, I went with this name because `update` looks a bit like something that lets you _edit_ stuff. It doesn't publish anything, either, that's for sure.

File: kudosaurus.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	PutKudo(teamID string, idFrom string, idTo string, text string)
```
Comment (TheMacies) [inline kudosaurus.go:40]: You do not have to write the type each time if consecutive variables are of the same type.
In this case

`PutKudo(teamID, idFrom, idTo, text string)`

would be valid and much shorter.

File: internal/server/actions.go (hunk 8 above)
Comment (Writhe) [reply internal/server/actions.go:44]: Really? I thought I need an explicit call to `WriteHeader` or `Write` to actually - you know - write a response. Note that what happens below this line is a separate request, not a response.

File: kudosaurus.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+	GetKudo(id string) (Kudo, error)
+	GetPerson(teamID string, id string, includeKudos bool) (Person, bool)
+	RemoveKudo(id string) error
+	GetKudos(teamID string) []Kudo
+	GetTeams() []Team
+	PutTeam(team Team)
+	GetTeam(id string) (Team, bool)
+	SetAdmins(teamID string, adminIDs []string)
+	GetAdmins(teamID string) []string
+	CheckAdmin(teamID string, userID string) (isAdmin bool, isFound bool)
```
Comment (TheMacies) [inline kudosaurus.go:50]: Each method in this interface should have an error variable as one of its returns. I can imagine that some remote SQL db dies and returns some valuable information of why it cannot handle your requests. Also, `isFound bool` could be replaced with error as well. You could return some concrete error (pattern called sentinel errors) in case of 'admin not found' scenario and check for it in your `if err != nil` clause. For example

```
isAdmin, err := store.CheckAdmin(tID,uID)
if err != nil {
    if errors.Is(err, kudosaurus.ErrAdminNotFound){
      //handle if not found
    }
    //handle if could not verify if is admin
} 
```

**Review by KNiepok — COMMENTED** [review]
A lot of work is being done here!
What the project needs is a better structure: moving the logic from API functions to services, which Pati proposed already.
Also many methods (like storage methods) should accept ctx as first argument and could return errors. 
For instance 
`func (s *SqliteStore) getPerson(teamID string, id string) (Person, bool) {
	var person Person
	s.db.First(&person, "team_slack_id = ? AND slack_id = ?", teamID, id)
	return person, person.SlackID != ""
}`
returns bool that indicates if person was found or not, but returning error may provide additional information (is the person missing? connection to DB is lost?)

File: internal/misc/bullets.go [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package misc
```
Comment (TheMacies) [inline internal/misc/bullets.go:1]: misc/utils/helpers/common packages are root of all evil 👿 
This function here depends on **slacks's** emotes, therefore should be in slack package.

The same for exclamation file.

File: internal/templates/home.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+import (
+	"fmt"
+	"strings"
+
+	"github.com/writhe/kudosaurus"
+)
+
+// Kudo - kudo view block
+func Kudo(kudo kudosaurus.Kudo) string {
+	template := `
```
Comment (TheMacies) [inline internal/templates/home.go:12]: This templates are tied directly to slack format - this all should be in slack package.

File: kudosaurus.go (hunk 9 above)
Comment (Writhe) [reply kudosaurus.go:39]: I'll have to look into the entire concept of contexts - I don't think it has a good equivalent in anything I used so far. Or I might not see the similarity.

File: internal/server/actions.go (hunk 8 above)
Comment (KNiepok) [reply internal/server/actions.go:44]: `// WriteHeader sends an HTTP response header with the provided
	// status code.
	//
	// If WriteHeader is not called explicitly, the first call to Write
	// will trigger an implicit WriteHeader(http.StatusOK).
	// Thus explicit calls to WriteHeader are mainly used to
	// send error codes.
	//
	// The provided code must be a valid HTTP 1xx-5xx status code.
	// Only one header may be written. Go does not currently
	// support sending user-defined 1xx informational headers,
	// with the exception of 100-continue response header that the
	// Server sends automatically when the Request.Body is read.
	WriteHeader(statusCode int)`

OK, now I see that you don't `Write` anywhere, so you actually need to  `WriteHeader`

File: kudosaurus.go (hunk 10 above)
Comment (Writhe) [reply kudosaurus.go:40]: Oooh. Handy.

File: kudosaurus.go (hunk 11 above)
Comment (Writhe) [reply kudosaurus.go:50]: Got it. 👍

File: internal/sqlitestore/team.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
+package sqlitestore
+
+import (
+	"github.com/jinzhu/gorm"
+	"github.com/writhe/kudosaurus"
+)
+
+// Team - person model
+type Team struct {
```
Comment (TheMacies) [inline internal/sqlitestore/team.go:9]: I don't see why this type has to be exported. This is some internal model of sqlite storage. I guess it could be made internal. Same goes for other model types.

File: internal/sqlitestore/team.go [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+	gorm.Model
+	SlackID  string
+	Name     string
+	People   []Person `gorm:"foreignkey:TeamSlackID;association_foreignkey:SlackID"`
+	Settings Settings `gorm:"foreignkey:TeamSlackID;association_foreignkey:SlackID"`
+}
+
+func (t *Team) getData() kudosaurus.Team {
+	people := []kudosaurus.Person{}
```
Comment (TheMacies) [inline internal/sqlitestore/team.go:18]: Just a small performance tip. You can preallocate such slices with 

`people := make([]kudosaurus.Person{},0,len(t.People))`

File: internal/sqlitestore/store.go (hunk 5 above)
Comment (KNiepok) [reply internal/sqlitestore/store.go:17]: The best part is that when you swap sqlite for sql or psql, you don't even have to change that code (only imports) 🏅

File: internal/misc/bullets.go (hunk 12 above)
Comment (Writhe) [reply internal/misc/bullets.go:1]: Yeah, I figured it's `misc` but in the scope of the app, not like "Global `misc`! Hey everybody, let's use it all over the place!". I think I see the flaw in this assumption, now.

File: internal/sqlitestore/store.go [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+
+func beginningOfMonth(t time.Time) int64 {
+	return time.Date(t.Year(), t.Month(), 1, 0, 0, 0, 0, t.Location()).Unix()
+}
+
+func beginningOfCurrentMonth() int64 {
+	return beginningOfMonth(time.Now())
+}
+
+func beginningOfPreviousMonth() int64 {
```
Comment (TheMacies) [inline internal/sqlitestore/store.go:116]: Is this wizardry even used :D ?

Comment (Writhe) [reply internal/sqlitestore/store.go:116]: Not anymore, it's not. I wasn't ready to part with it, yet. ;)

File: internal/sqlitestore/store.go (hunk 5 above)
Comment (TheMacies) [reply internal/sqlitestore/store.go:17]: you actually have to change this line 👉  `db, err := gorm.Open("sqlite3", s.path)`

File: kudosaurus.go (hunk 9 above)
Comment (TheMacies) [reply kudosaurus.go:39]: I do not think there is actually. And also, many different usages of context are very opinionated. Although, using context as a circuit breaker is generally a way to go and it's very handy.

File: cmd/main.go [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/writhe/kudosaurus/internal/sqlitestore"
+)
+
+func main() {
+	logger := log.New(os.Stdout, "", log.Ltime|log.Ldate|log.LUTC)
+
+	cfg, err := config.GetConfig("config.yml")
+
+	if err != nil {
+		panic(err)
```
Comment (TheMacies) [inline cmd/main.go:18]: This should be more descriptive. Also, you could use `log.Fatal` since it closes your program as well, however without printing the unnecessary stacktrace :).

File: internal/config/config.go [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+	f, err := os.Open(path)
+	if err != nil {
+		return Config{}, fmt.Errorf("could not open config file - '%s'", path)
+	}
+	defer f.Close()
+
+	var cfg Config
+	decoder := yaml.NewDecoder(f)
+	err = decoder.Decode(&cfg)
+	if err != nil {
```
Comment (TheMacies) [inline internal/config/config.go:43]: These could be merged into one line.
`if err := yaml.NewDecoder(f).Decode(&cfg); err != nil {`

File: internal/config/config.go [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+		MaxKudos int `yaml:"maxKudos" validate:"required"`
+	} `yaml:"settings" validate:"required"`
+}
+
+// GetConfig - gets config from file
+func GetConfig(path string) (Config, error) {
+	validate = validator.New()
+	f, err := os.Open(path)
+	if err != nil {
+		return Config{}, fmt.Errorf("could not open config file - '%s'", path)
```
Comment (TheMacies) [inline internal/config/config.go:36]: You should be wrapping errors that are initial causes instead of creating new ones.

`return Config{}, fmt.Errorf("could not open config file: %w", err)`

File: internal/config/config.go [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+	}
+	defer f.Close()
```
Comment (TheMacies) [inline internal/config/config.go:38]: handling files is always nasty :)
Have a look at ioutil.ReadFile - you wouldn't have to close this boy here

File: internal/config/config.go [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+
+	var cfg Config
+	decoder := yaml.NewDecoder(f)
+	err = decoder.Decode(&cfg)
+	if err != nil {
+		return Config{}, fmt.Errorf("could not decode the config: %s", err.Error())
+	}
+	err = validate.Struct(cfg)
+	if err != nil {
+		return Config{}, fmt.Errorf("could not validate the config: %s", err.Error())
```
Comment (TheMacies) [inline internal/config/config.go:48]: You should never do that - do not create new errors with injected strings from the underlying errors. Errors must always be wrapped like this: `fmt.Errorf("could not validate the config: %w", err) ` or for older version of go `errors.Wrap(err, "could not (...))`. This allows you to do some further assertions on concrete implementations of Error interface.

File: internal/server/actions.go [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+
+func (s *CommandServer) handleAction(res http.ResponseWriter, req *http.Request) {
+	if err := s.verifySigningSecret(req, res); err != nil {
+		return
+	}
+
+	var payload slack.InteractionCallback
+
+	err := json.Unmarshal([]byte(req.FormValue("payload")), &payload)
+
```
Comment (TheMacies) [inline internal/server/actions.go:135]: this newline here hurts 🌵

File: internal/server/commands.go [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+
+	"github.com/slack-go/slack"
+	"github.com/writhe/kudosaurus/internal/misc"
+)
+
+func (s *CommandServer) handleKudo(c slack.SlashCommand, w http.ResponseWriter) {
+	parts := strings.Fields(c.Text)
+
+	userToken := parts[0]
+	re := regexp.MustCompile(`<@(([A-Z0-9])+)(\|([a-z0-9\.]){0,21})?>`)
```
Comment (TheMacies) [inline internal/server/commands.go:17]: This could be moved to a package-level variable so you wouldn't have to parse the regexp each time.

File: internal/server/commands.go [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+
+	targetUserID := targetUserMatches[1]
+	targetUserName, foundTargetUser := s.getUserName(targetUserID)
+
+	if !foundTargetUser {
+		fmt.Println("No target user")
+		return
+	}
+
+	body := strings.Join(parts[1:], " ")
```
Comment (TheMacies) [inline internal/server/commands.go:32]: potential panic here ? Some assertion for len of `parts` is required

File: internal/server/commands.go [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+	)
+}
+
+func (s *CommandServer) handleCommand(res http.ResponseWriter, req *http.Request) {
+	if err := s.verifySigningSecret(req, res); err != nil {
+		return
+	}
+
+	c, err := slack.SlashCommandParse(req)
+
```
Comment (TheMacies) [inline internal/server/commands.go:79]: 🌵

File: internal/server/server.go [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+		return s.closeWithError(err, http.StatusUnauthorized, res)
+	}
+
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return s.closeWithError(err, http.StatusInternalServerError, res)
+	}
+	req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+	verifier.Write(body)
```
Comment (TheMacies) [inline internal/server/server.go:86]: seems to me that you could do `io.Copy(verifier, req.Body)` instead

Comment (TheMacies) [reply internal/server/server.go:86]: Oh, you also want to have a copy of it. Then do 
```
bodyCopy := &bytes.Buffer{}
io.Copy(verifier, io.TeeReader(req.Body, bodyCopy))
```

File: internal/server/server.go [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+	return nil
+}
+
+func (s *CommandServer) sendJSON(payload interface{}, res http.ResponseWriter) {
+	b, err := json.Marshal(payload)
+	if err != nil {
+		s.closeWithError(err, http.StatusInternalServerError, res)
+	}
+	res.Header().Set("Content-Type", "application/json")
+	res.Write(b)
```
Comment (TheMacies) [inline internal/server/server.go:100]: Also simplifiable - 
```
res.Header().Set("Content-Type", "application/json")
json.NewEncoder(res).Encode(payload)
```

File: internal/sqlitestore/store.go [hunk 28]
```diff
@@ -0,0 +1,246 @@
...
+	return result, nil
+}
+
+// PutKudo puts kudo
+func (s *SqliteStore) PutKudo(teamID string, idFrom string, idTo string, text string) {
+	author, authorFound := s.getPerson(teamID, idFrom)
+	owner, ownerFound := s.getPerson(teamID, idTo)
+
+	if !(authorFound && ownerFound) {
+		s.logger.Printf("No such id '%s'.\n", idTo)
```
Comment (TheMacies) [inline internal/sqlitestore/store.go:73]: Misleading print,  `idFrom` may be wrong as well
